x/launch/keeper/msg_server_settle_request_test.go

```diff
@@ -0,0 +1,268 @@
+			},
+			err: sdkerrors.Wrap(types.ErrChainNotFound, invalidChain),
+		}, {
+			name: "launch triggered chain",
+			msg: types.MsgSettleRequest{
+				ChainID:     chains[0].ChainID,
+				Coordinator: addr1,
+				RequestID:   requests[0].RequestID,
+			},
+			err: sdkerrors.Wrap(types.ErrTriggeredLaunch, addr1),
+		}, {
+			name: "no permission error",
+			msg: types.MsgSettleRequest{
+				ChainID:     chains[1].ChainID,
+				Coordinator: addr1,
+				RequestID:   requests[0].RequestID,
+			},
+			err: sdkerrors.Wrap(types.ErrNoAddressPermission, addr1),
+		}, {
+			name: "settle a invalid request",
+			msg: types.MsgSettleRequest{
+				ChainID:     chains[1].ChainID,
+				Coordinator: pk.GetCoordinatorAddressFromID(sdkCtx, chains[1].CoordinatorID),
+				RequestID:   99999999,
+			},
+			err: sdkerrors.Wrap(types.ErrRequestNotFound,
+				"request 999999 for chain %s not found"),
+		}, {
+			name: "add chain 1 request 1",
+			msg: types.MsgSettleRequest{
+				ChainID:     chains[1].ChainID,
+				Coordinator: pk.GetCoordinatorAddressFromID(sdkCtx, chains[1].CoordinatorID),
+				RequestID:   requests[0].RequestID,
+			},
+			want: 0,
+		}, {
+			name: "add chain 1 request 2",
+			msg: types.MsgSettleRequest{
+				ChainID:     chains[1].ChainID,
+				Coordinator: pk.GetCoordinatorAddressFromID(sdkCtx, chains[1].CoordinatorID),
+				RequestID:   requests[1].RequestID,
+			},
+			want: 0,
+		}, {
+			name: "add chain 1 request 3",
+			msg: types.MsgSettleRequest{
+				ChainID:     chains[1].ChainID,
+				Coordinator: pk.GetCoordinatorAddressFromID(sdkCtx, chains[1].CoordinatorID),
+				RequestID:   requests[2].RequestID,
+			},
+			want: 1,
+		}, {
+			name: "add chain 2 request 5",
+			msg: types.MsgSettleRequest{
+				ChainID:     chains[1].ChainID,
+				Coordinator: pk.GetCoordinatorAddressFromID(sdkCtx, chains[1].CoordinatorID),
+				RequestID:   requests[5].RequestID,
+			},
+			want: 0,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := srv.SettleRequest(ctx, &tt.msg)
+			if tt.err != nil {
+				require.ErrorIs(t, err, tt.err)
+				return
+			}
+			require.NoError(t, err)
+
+			_, found := k.GetRequest(sdkCtx, tt.msg.ChainID, tt.msg.RequestID)
+			require.False(t, found, "request not removed")
+		})
+	}
+}
+
+func TestApplyRequest(t *testing.T) {
+	var (
+		genesis               = sample.AccAddress()
+		vested                = sample.AccAddress()
+		validator             = sample.AccAddress()
+		k, _, _, _, sdkCtx, _ = setupMsgServer(t)
+		chainID, _            = sample.ChainID(10)
+		contents              = sample.CreateAllRequestContents(chainID, genesis, vested, validator)
+		requests              = createRequests(k, sdkCtx, chainID, contents)
+		invalidContent, _     = codec.NewAnyWithValue(&types.Request{})
+		invalidContentID      = k.AppendRequest(sdkCtx, *sample.RequestWithContent(chainID, invalidContent))
+	)
+	tests := []struct {
+		name string
+		msg  types.MsgSettleRequest
```

Comment:
Can we have more generic tests for `applyRequest` where `types.Request` is directly given in the test case